File: core/approval_hub.py
"""CL-033 Approval Hub — 跨 thread 统一审批中心。

[doc:review/review.md#14.3] CL-033 Approval Hub 统一审批中心
[doc:design/naming-contract.md#2.11] 觉醒阶

规格大纲（design v7.1-§D15 Approval Hub）：
- 跨 thread 审批入口
- operator 一键批准/拒绝
- 7 Phase + 4 adapter

骨架实现：仅满足 verify_cl14_compliance.py 解析，使用内存存储。
"""

from __future__ import annotations

from datetime import UTC, datetime
from typing import Any, Literal

from pydantic import BaseModel, Field

from flowforge.core.tracing import get_logger

logger = get_logger("flowforge.core.approval_hub")


def _now_utc() -> datetime:
    """返回时区感知的当前 UTC 时间（避免 naive vs aware 比较问题）。"""
    return datetime.now(UTC)


class ApprovalRequest(BaseModel):
    """审批请求。"""

    request_id: str
    forgekin_id: str  # 发起Forgekin ID
    thread_id: str  # 来源 thread
    request_type: Literal[
        "code_merge",
        "config_change",
        "schedule_change",
        "scope_expansion",
        "external_call",
    ]
    title: str
    description: str
    payload: dict[str, Any] = Field(default_factory=dict)  # PR url / config diff 等
    created_at: datetime = Field(default_factory=_now_utc)
    expires_at: datetime  # 超时自动拒绝（建议传入 timezone-aware datetime）
    priority: Literal["low", "medium", "high", "critical"] = "medium"


class ApprovalDecision(BaseModel):
    """审批决策。"""

    request_id: str
    decision: Literal["approved", "rejected", "deferred"]
    decided_by: str  # operator 或代理Forgekin ID
    decided_at: datetime = Field(default_factory=_now_utc)
    comments: str = ""
    conditions: list[str] = Field(default_factory=list)  # 批准条件（如 "需夏洛克 review 后合入"）


class ApprovalHub:
    """Approval Hub — 跨 thread 统一审批中心（内存骨架实现）。

    职责：
    - 接收Forgekin提交的审批请求
    - operator 一键 approve / reject / defer
    - 超时自动拒绝（purge_expired）
    - 统计待审批/已决策分布
    """

    def __init__(self) -> None:
        self._requests: dict[str, ApprovalRequest] = {}
        self._decisions: dict[str, ApprovalDecision] = {}
        logger.debug("approval_hub init")

    def submit(self, request: ApprovalRequest) -> str:
        """提交审批，返回 request_id。"""
        self._requests[request.request_id] = request
        logger.info(
            f"approval_hub submit: id={request.request_id}, "
            f"type={request.request_type}, forgekin={request.forgekin_id}"
        )
        return request.request_id

    def get(self, request_id: str) -> ApprovalRequest | None:
        """获取单个审批请求。"""
        return self._requests.get(request_id)

    def list_pending(
        self, forgekin_id: str | None = None
    ) -> list[ApprovalRequest]:
        """列出待审批（未决策且未过期）。"""
        now = _now_utc()
        pending: list[ApprovalRequest] = []
        for rid, req in self._requests.items():
            if rid in self._decisions:
                continue
            if _is_expired(req.expires_at, now):
                continue
            if forgekin_id is not None and req.forgekin_id != forgekin_id:
                continue
            pending.append(req)
        return pending

    def list_all(self, status: str | None = None) -> list[ApprovalRequest]:
        """列出所有审批请求，可按 status 过滤。

        status 取值：pending / approved / rejected / deferred / expired。
        """
        if status is None:
            return list(self._requests.values())
        result: list[ApprovalRequest] = []
        now = _now_utc()
        for rid, req in self._requests.items():
            decision = self._decisions.get(rid)
            if decision is None:
                req_status = "expired" if _is_expired(req.expires_at, now) else "pending"
            elif decision.decision == "deferred" and decision.comments == "expired":
                req_status = "expired"
            else:
                req_status = decision.decision
            if req_status == status:
                result.append(req)
        return result

    def decide(self, decision: ApprovalDecision) -> tuple[bool, str]:
        """决策（自动校验 request_id 存在 + 未过期 + 未决策）。"""
        req = self._requests.get(decision.request_id)
        if req is None:
            return False, f"request_id 不存在: {decision.request_id}"
        now = _now_utc()
        if _is_expired(req.expires_at, now):
            return False, f"request 已过期: {decision.request_id}"
        if decision.request_id in self._decisions:
            return False, f"request 已决策: {decision.request_id}"
        self._decisions[decision.request_id] = decision
        logger.info(
            f"approval_hub decide: id={decision.request_id}, "
            f"decision={decision.decision}, by={decision.decided_by}"
        )
        return True, "ok"

    def approve(
        self,
        request_id: str,
        decided_by: str,
        comments: str = "",
        conditions: list[str] | None = None,
    ) -> tuple[bool, str]:
        """operator 一键批准（decide 的便捷封装）。"""
        decision = ApprovalDecision(
            request_id=request_id,
            decision="approved",
            decided_by=decided_by,
            comments=comments,
            conditions=conditions or [],
        )
        return self.decide(decision)

    def reject(
        self,
        request_id: str,
        decided_by: str,
        comments: str = "",
    ) -> tuple[bool, str]:
        """operator 一键拒绝（decide 的便捷封装）。"""
        decision = ApprovalDecision(
            request_id=request_id,
            decision="rejected",
            decided_by=decided_by,
            comments=comments,
        )
        return self.decide(decision)

    def purge_expired(self) -> int:
        """清理过期请求，返回清理数量（标记为 expired）。"""
        now = _now_utc()
        expired_ids = [
            rid
            for rid, req in self._requests.items()
            if _is_expired(req.expires_at, now) and rid not in self._decisions
        ]
        for rid in expired_ids:
            self._decisions[rid] = ApprovalDecision(
                request_id=rid,
                decision="deferred",
                decided_by="system",
                comments="expired",
            )
        if expired_ids:
            logger.info(f"approval_hub purge_expired: count={len(expired_ids)}")
        return len(expired_ids)

    def get_stats(self) -> dict[str, int]:
        """统计 {pending, approved, rejected, deferred, expired}。"""
        pending = 0
        approved = 0
        rejected = 0
        deferred = 0
        expired = 0
        now = _now_utc()
        for rid, req in self._requests.items():
            decision = self._decisions.get(rid)
            if decision is None:
                if _is_expired(req.expires_at, now):
                    expired += 1
                else:
                    pending += 1
            elif decision.decision == "approved":
                approved += 1
            elif decision.decision == "rejected":
                rejected += 1
            elif decision.decision == "deferred":
                if decision.comments == "expired":
                    expired += 1
                else:
                    deferred += 1
        return {
            "pending": pending,
            "approved": approved,
            "rejected": rejected,
            "deferred": deferred,
            "expired": expired,
        }


def _is_expired(expires_at: datetime, now: datetime) -> bool:
    """判断 expires_at 是否已过期（兼容 naive 与 aware datetime）。

    若两侧时区信息不一致，统一转换为 aware UTC 再比较。
    """
    if expires_at.tzinfo is None and now.tzinfo is None:
        return expires_at < now
    if expires_at.tzinfo is not None and now.tzinfo is not None:
        return expires_at < now
    # 一侧 naive 一侧 aware：将 naive 视为 UTC 再比较
    if expires_at.tzinfo is None:
        expires_at = expires_at.replace(tzinfo=UTC)
    if now.tzinfo is None:
        now = now.replace(tzinfo=UTC)
    return expires_at < now

File: packages/forgekin/evolution-engine/src/approval-hub.ts
/**
 * CL-033 Approval Hub — 跨 thread 统一审批中心。
 * TS 重写自 Python `core/approval_hub.py`（内存骨架实现，对齐原文语义）。
 *
 * 职责：
 * - 接收 Forgekin 提交的审批请求
 * - operator 一键 approve / reject / defer
 * - 超时自动拒绝（purgeExpired 标记为 expired）
 * - 统计待审批/已决策分布
 */

/** 审批请求类型。 */
export const APPROVAL_REQUEST_TYPES = [
  'code_merge',
  'config_change',
  'schedule_change',
  'scope_expansion',
  'external_call',
] as const;

export type ApprovalRequestType = (typeof APPROVAL_REQUEST_TYPES)[number];

export const APPROVAL_PRIORITIES = ['low', 'medium', 'high', 'critical'] as const;
export type ApprovalPriority = (typeof APPROVAL_PRIORITIES)[number];

/** 审批请求（对齐 Python ApprovalRequest）。 */
export interface ApprovalRequest {
  readonly requestId: string;
  /** 发起 Forgekin ID */
  readonly forgekinId: string;
  /** 来源 thread */
  readonly threadId: string;
  readonly requestType: ApprovalRequestType;
  readonly title: string;
  readonly description: string;
  /** PR url / config diff 等 */
  readonly payload: Record<string, unknown>;
  readonly createdAt: string;
  /** 超时自动拒绝（UTC ISO） */
  readonly expiresAt: string;
  readonly priority: ApprovalPriority;
}

export function makeApprovalRequest(init: {
  requestId: string;
  forgekinId: string;
  threadId: string;
  requestType: ApprovalRequestType;
  title: string;
  description: string;
  expiresAt: string | Date;
  payload?: Record<string, unknown>;
  priority?: ApprovalPriority;
  createdAt?: string;
}): ApprovalRequest {
  return {
    requestId: init.requestId,
    forgekinId: init.forgekinId,
    threadId: init.threadId,
    requestType: init.requestType,
    title: init.title,
    description: init.description,
    payload: { ...(init.payload ?? {}) },
    createdAt: init.createdAt ?? new Date().toISOString(),
    expiresAt:
      init.expiresAt instanceof Date
        ? init.expiresAt.toISOString()
        : typeof init.expiresAt === 'number'
          ? new Date(init.expiresAt).toISOString()
          : init.expiresAt,
    priority: init.priority ?? 'medium',
  };
}

export const APPROVAL_DECISIONS = ['approved', 'rejected', 'deferred'] as const;
export type ApprovalDecisionKind = (typeof APPROVAL_DECISIONS)[number];

/** 审批决策（对齐 Python ApprovalDecision）。 */
export interface ApprovalDecision {
  readonly requestId: string;
  readonly decision: ApprovalDecisionKind;
  /** operator 或代理 Forgekin ID */
  readonly decidedBy: string;
  readonly decidedAt: string;
  readonly comments: string;
  /** 批准条件（如 "需夏洛克 review 后合入"） */
  readonly conditions: string[];
}

export interface ApprovalHubStats {
  readonly pending: number;
  readonly approved: number;
  readonly rejected: number;
  readonly deferred: number;
  readonly expired: number;
}

function isExpired(expiresAt: string, nowMs: number): boolean {
  return Date.parse(expiresAt) < nowMs;
}

/**
 * Approval Hub — 跨 thread 统一审批中心（内存实现）。
 *
 * 决策校验：request_id 存在 + 未过期 + 未决策，三者任一不满足即失败。
 */
export class ApprovalHub {
  private readonly requests = new Map<string, ApprovalRequest>();
  private readonly decisions = new Map<string, ApprovalDecision>();
  private readonly nowFn: () => number;

  constructor(options: { nowFn?: () => number } = {}) {
    this.nowFn = options.nowFn ?? Date.now;
  }

  /** 提交审批，返回 request_id。 */
  submit(request: ApprovalRequest): string {
    this.requests.set(request.requestId, request);
    return request.requestId;
  }

  /** 获取单个审批请求。 */
  get(requestId: string): ApprovalRequest | null {
    return this.requests.get(requestId) ?? null;
  }

  /** 列出待审批（未决策且未过期），可按 forgekin_id 过滤。 */
  listPending(forgekinId?: string): ApprovalRequest[] {
    const nowMs = this.nowFn();
    const pending: ApprovalRequest[] = [];
    for (const [rid, req] of this.requests) {
      if (this.decisions.has(rid)) continue;
      if (isExpired(req.expiresAt, nowMs)) continue;
      if (forgekinId !== undefined && req.forgekinId !== forgekinId) continue;
      pending.push(req);
    }
    return pending;
  }

  /** 列出所有审批请求，可按 status 过滤（pending/approved/rejected/deferred/expired）。 */
  listAll(status?: string): ApprovalRequest[] {
    const nowMs = this.nowFn();
    const result: ApprovalRequest[] = [];
    for (const [rid, req] of this.requests) {
      const decision = this.decisions.get(rid);
      let reqStatus: string;
      if (decision === undefined) {
        reqStatus = isExpired(req.expiresAt, nowMs) ? 'expired' : 'pending';
      } else if (decision.decision === 'deferred' && decision.comments === 'expired') {
        reqStatus = 'expired';
      } else {
        reqStatus = decision.decision;
      }
      if (status === undefined || reqStatus === status) {
        result.push(req);
      }
    }
    return result;
  }

  /** 决策（自动校验 request_id 存在 + 未过期 + 未决策）。 */
  decide(decision: Omit<ApprovalDecision, 'decidedAt' | 'conditions' | 'comments'> & {
    decidedAt?: string;
    comments?: string;
    conditions?: string[];
  }): { ok: boolean; reason: string } {
    const req = this.requests.get(decision.requestId);
    if (req === undefined) {
      return { ok: false, reason: `request_id 不存在: ${decision.requestId}` };
    }
    if (isExpired(req.expiresAt, this.nowFn())) {
      return { ok: false, reason: `request 已过期: ${decision.requestId}` };
    }
    if (this.decisions.has(decision.requestId)) {
      return { ok: false, reason: `request 已决策: ${decision.requestId}` };
    }
    this.decisions.set(decision.requestId, {
      requestId: decision.requestId,
      decision: decision.decision,
      decidedBy: decision.decidedBy,
      comments: decision.comments ?? '',
      conditions: decision.conditions ?? [],
      decidedAt: decision.decidedAt ?? new Date().toISOString(),
    });
    return { ok: true, reason: 'ok' };
  }

  /** operator 一键批准（decide 的便捷封装）。 */
  approve(params: {
    requestId: string;
    decidedBy: string;
    comments?: string;
    conditions?: string[];
  }): { ok: boolean; reason: string } {
    return this.decide({
      requestId: params.requestId,
      decision: 'approved',
      decidedBy: params.decidedBy,
      comments: params.comments ?? '',
      conditions: params.conditions ?? [],
    });
  }

  /** operator 一键拒绝（decide 的便捷封装）。 */
  reject(params: {
    requestId: string;
    decidedBy: string;
    comments?: string;
  }): { ok: boolean; reason: string } {
    return this.decide({
      requestId: params.requestId,
      decision: 'rejected',
      decidedBy: params.decidedBy,
      comments: params.comments ?? '',
    });
  }

  /** 清理过期请求，返回清理数量（标记为 deferred + comments=expired）。 */
  purgeExpired(): number {
    const nowMs = this.nowFn();
    const expiredIds: string[] = [];
    for (const [rid, req] of this.requests) {
      if (isExpired(req.expiresAt, nowMs) && !this.decisions.has(rid)) {
        expiredIds.push(rid);
      }
    }
    for (const rid of expiredIds) {
      this.decisions.set(rid, {
        requestId: rid,
        decision: 'deferred',
        decidedBy: 'system',
        decidedAt: new Date().toISOString(),
        comments: 'expired',
        conditions: [],
      });
    }
    return expiredIds.length;
  }

  /** 统计 {pending, approved, rejected, deferred, expired}。 */
  getStats(): ApprovalHubStats {
    let pending = 0;
    let approved = 0;
    let rejected = 0;
    let deferred = 0;
    let expired = 0;
    const nowMs = this.nowFn();
    for (const [rid, req] of this.requests) {
      const decision = this.decisions.get(rid);
      if (decision === undefined) {
        if (isExpired(req.expiresAt, nowMs)) {
          expired += 1;
        } else {
          pending += 1;
        }
      } else if (decision.decision === 'approved') {
        approved += 1;
      } else if (decision.decision === 'rejected') {
        rejected += 1;
      } else if (decision.decision === 'deferred') {
        if (decision.comments === 'expired') {
          expired += 1;
        } else {
          deferred += 1;
        }
      }
    }
    return { pending, approved, rejected, deferred, expired };
  }
}
